src/Education/XamlGenerator/IRuleHelpXamlBuilder.cs

```diff
@@ -0,0 +1,472 @@
+
+        /// <summary>
+        /// Stack of currently open XAML elements
+        /// </summary>
+        /// <remarks>We need some information about the current structure so we can check whether some 
+        /// operations are valid e.g. can be we add text to the current element?</remarks>
+        private Stack<XamlOutputElementInfo> outputXamlElementStack = new Stack<XamlOutputElementInfo>();
+
+        /// <summary>
+        /// Used to add background colour to alternate rows in a table
+        /// </summary>
+        private bool tableAlternateRow;
+
+        public FlowDocument Create(IRuleHelp ruleHelp)
+        {
+            var xaml = CreateXamlString(ruleHelp);
+
+            using (var stringReader = new StringReader(xaml))
+            {
+                var xamlReader = XmlReader.Create(stringReader);
+                var flowDocument = (FlowDocument)XamlReader.Load(xamlReader);
+                return flowDocument;
+            }
+        }
+
+        internal /* for testing */ string CreateXamlString(IRuleHelp ruleHelp)
+        {
+            var sb = new StringBuilder();
+            writer = CreateXmlWriter(sb);
+            reader = CreateXmlReader(ruleHelp.HtmlDescription);
+
+            try
+            {
+                WriteDocumentHeader(ruleHelp);
+
+                while (reader.Read())
+                {
+                    switch (reader.NodeType)
+                    {
+                        case XmlNodeType.Element:
+                            ProcessElement();
+                            break;
+
+                        case XmlNodeType.Text:
+                            WriteText(reader.Value);
+                            break;
+
+                        case XmlNodeType.Whitespace:
+                            writer.WriteString(reader.Value);
+                            break;
+
+                        case XmlNodeType.EndElement:
+                            XamlOutputElementInfo xamlOutputElement;
+                            do
+                            {
+                                xamlOutputElement = outputXamlElementStack.Pop();
+                                writer.WriteEndElement();
+                            } while (xamlOutputElement.HtmlElementName != reader.Name);
+
+                            break;
+                        default:
+                            var message = string.Format(Resources.XamlBuilder_UnexpectedNodeError, reader.NodeType, reader.Name, reader.Value);
+                            throw new InvalidDataException(message);
+                    }
+                }
+
+                // We've processed all of the html elements.
+                // Now, the stack should only contain the root FlowDocument element, plus an extra
+                // block element if the first tag we processed was an Inline.
+                Debug.Assert(outputXamlElementStack.Count <= 2, "Expecting at most 2 unclosed elements in the stack");
```

Comment:
Sanity check when we've finished writing that we have the expected items left on the stack.